docs/docs-beta/docs/guides/io-managers.md

```diff
@@ -4,43 +4,85 @@ sidebar_position: 50
+- You have assets that load upstream dependencies into memory to do the computation
 
-Using an I/O manager can help simplify the code responsible for reading and writing the data to and from a data source for assets that follow this pattern.
+**I/O managers may not be the best fit if:**
 
-For example, both assets in the code below are constructing a DuckDB connection object, reading from an upstream table, applying some in-memory transform, and then writing the transformed result into a new table in DuckDB.
+- You want to run SQL queries that create or update a table in a database
+- Your pipeline manages I/O on its own by using other libraries/tools that write to storage
+- Your assets won't fit in memory, such as a database table with billions of rows
 
-<CodeExample filePath="guides/external-systems/assets-without-io-managers.py" language="python" title="Assets without I/O managers" />
+As a general rule, if your pipeline becomes more complicated in order to use I/O managers, it's likely that I/O managers aren't a good fit. In these cases you should use `deps` to [define dependencies](/guides/asset-dependencies).
 
-To switch to using I/O managers, we can use the DuckDB / Pandas I/O manager provided by the  `dagster_duckdb_pandas` package.  The I/O manager will be used to read and write data instead of the `DuckDBResource`.
+## Using I/O managers in assets \{#io-in-assets}
 
-To load an upstream asset using an I/O manager, that asset is specified as an input parameter to the asset function rather than within the `deps` list on the `@asset` decorator.  The `DuckDBPandasIOManager` reads the DuckDB table with the same name as the upstream asset and passes the data into the asset function as a Pandas DataFrame.
+Consider the following example, which contains assets that construct a DuckDB connection object, read data from an upstream table, apply some in-memory transform, and write the result to a new table in DuckDB:
 
-To store data using an I/O manager, return the data in the asset function. The data returned must be a valid type. The `DuckDBPandasIOManager` accepts Pandas data frames and writes them to DuckDB as a table with the same name as the asset.
+<CodeExample filePath="guides/external-systems/assets-without-io-managers.py" language="python" />
+
+Using an I/O manager would remove the code that reads and writes data from the assets themselves, instead delegating it to the I/O manager. The assets would be left only with the code that applies transformations or retrieves the initial CSV file.
+
+<CodeExample filePath="guides/external-systems/assets-with-io-managers.py" language="python" />
+
+To load upstream assets using an I/O manager, specify the asset as an input parameter to the asset function. In this example, the `DuckDBPandasIOManager` I/O manager will read the DuckDB table with the same name as the upstream asset (`raw_sales_data`) and pass the data to `clean_sales_data` as a Pandas DataFrame.
```

Comment:
I kept these as paragraphs instead of code comments since it felt like a lot. Thoughts?